src/agent/internals/src/main/java/com/google/devtools/cdbg/debuglets/java/FirebaseClient.java

```diff
@@ -0,0 +1,1038 @@
+   * @return the debuggee ID, in string form.
+   */
+  public String computeDebuggeeId(Debuggee debuggee) throws NoSuchAlgorithmException {
+    MessageDigest hash = MessageDigest.getInstance("SHA1");
+
+    // Ideally we could just generate a Json string and then hash that, however with Gson it seems
+    // there's no easy way to ensure the keys in the maps get output in a deterministic order.
+    updateHash(hash, GSON.toJsonTree(debuggee));
+    return "d-" + DatatypeConverter.printHexBinary(hash.digest()).substring(0, 8).toLowerCase();
+  }
+
+  /**
+   * Updates the current hash with the given JSON element. This is a recursive call, and for all map
+   * object will iterate over the keys/json fields in order to ensure the same hash will be computed
+   * for the same JSON content.
+   *
+   * @param hash the current hash to add to
+   * @param element the json element to add to the hash.
+   */
+  private static void updateHash(MessageDigest hash, JsonElement element) {
+    if (element.isJsonArray()) {
+      for (JsonElement e : element.getAsJsonArray()) {
+        updateHash(hash, e);
+      }
+    } else if (element.isJsonObject()) {
+      TreeMap<String, JsonElement> sortedElements = new TreeMap<>();
+      for (Map.Entry<String, JsonElement> kv : element.getAsJsonObject().entrySet()) {
+        sortedElements.put(kv.getKey(), kv.getValue());
+      }
+      for (Map.Entry<String, JsonElement> entry : sortedElements.entrySet()) {
+        updateHash(hash, entry.getValue());
+      }
+    } else if (!element.isJsonNull()) {
+      hash.update(element.toString().getBytes(UTF_8));
+    }
+  }
+
+  /**
+   * Configures a FirebaseApp instance and verifies the Firebase RTDB URL. On success, the value
+   * this.firebaseApp will be correctly set and the method returns normally, throws an Exception on
+   * error. Visible for testing
+   *
+   * @throws Exception when an error occurs or no valid database URL could be found.
+   */
+  void initializeFirebaseApp() throws Exception {
+    // Once we've successfully called FirebaseApp.initializeApp, there's no need to call it again.
+    // Also unless we call delete on the FirebaseApp, it will throw an exception.
+    if (firebaseApp != null) {
+      return;
+    }
+
+    ArrayList<String> urls = new ArrayList<>();
+    String configuredUrl = GcpEnvironment.getFirebaseDatabaseUrlString();
+    if (!configuredUrl.isEmpty()) {
+      urls.add(configuredUrl);
+    } else {
+      urls.add("https://" + metadata.getProjectId() + "-cdbg.firebaseio.com");
+      urls.add("https://" + metadata.getProjectId() + "-default-rtdb.firebaseio.com");
+    }
+
+    for (String databaseUrl : urls) {
+      this.firebaseApp = initializeFirebaseAppForUrl(databaseUrl);
+      if (this.firebaseApp != null) {
+        break;
+      }
+    }
+
+    if (this.firebaseApp == null) {
+      String errorMessage = "Failed to initialize FirebaseApp, attempted URls: " + urls;
```

Comment:
Odd capitalization on URls.  Also in the related test.

Reply:
Good catch, thanks.